### DISCUSSION TITLE
URLPattern Regex Capture Groups

### DISCUSSION THREAD

**Discussion by arnold-c** [gdh1995/vimium-c#1199] [Q&A] [answered]
I'm trying to write a search engine that allows me to look up directions using a separator (e.g. `:`) to delineate start and end locations. For example:

`gmdir New York City : Pittsburgh`

would be inserted into `https://www.google.com/maps/dir/New+York+City/Pittsburgh`

The regex `re=/(.+)\s:\s(.+)` would create the two capture groups correctly (from quick testing), but I'm unsure how this could be piped into the URL pattern. I tried adding this to the end of my search i.e.,

`gmdir: https://www.google.com/maps/dir/$s{$1}/$s{$2} [re=/(.+)\s:\s(.+)] Google Maps Directions`

but that didn't work, so I think I'm misunderstanding how VimiumC handles regex. If anyone has some suggestions, they would be very much welcomed!

**Comment by gdh1995** [comment, marked as answer]
1. The `[` and `]` just mean `re=` is optional.
2. sorry currently the solution for your issue does exist but is very very tricky (I'll try improving it in v2.12-15 in several months):

So please try:
1. add such a custom search engine
```
gmdir: vimium://sed/_gmdir\ https://www.google.com/maps/dir/$s{+} \
  https://www.google.com/maps/dir/ re=/^([\w+]+)\/([\w]+)/
  Google Maps Directions
```
2. add "text substitution rules" into "Auto substitution of various text":
```
_gmdir@\+(:|%3A)\+@/@i,return
_gmdir@\+([^/+]+)$@/$1@
```

Explain:
- the main search url is `vimium://sed/<sed_key>\ <url>`, which executes text substitions defined in `Auto substitution of various text`
- the prefix and `re=` will be used by commands like `searchInAnother` and `Vomnibar.activateUrl` to parse tab URL and get "search words"
  - the URL prefix before `re=` is a "prefix matcher" to filter out google maps urls
- `_gmdir` sed rules will check `+:+` firstly, and if no matches then replace the last `+word` with `/word`

**Reply by arnold-c** [reply]
Not sure how I missed this, but thank you! This works perfectly
